man/publish_content_views.8
.\" Manpage for sat_import.
.\" Contact ggatward@redhat.com to correct errors or typos.
.TH PUBLISH_CONTENT_VIEWS 8 "04 Jan 2017" "sat6_scripts" "sat6_scripts User Manual" man page"
.SH NAME
publish_content_views \- publish content views defined within Satellite 6

.SH SYNOPSIS
.B publish_content_views [\-o ORGANISATION] [\-a] [\-d] [\-l]
.LP
.B "publish_content_views --help"

.SH DESCRIPTION
.B publish_content_views
is a custom python script for publishing defined content views within a Satellite 6 server. The content must have been imported and synchronised prior to the publish being performed.


.SH OPTIONS
The options that apply to the
.B publish_content_views
command are:
.PP
.BR "-o", " --org"
.I "ORGANISATION"
.RS 3
The organisation within Satellite to perform the activity for. The default organisation is defined in the
.I config.yml
file, but can be overridden with this option.
.RE
.PP
.BR "-a", " --all"
.RS 3
Publish all configured content views. By default
.B publish_content_views
will publish only the content views defined in the
.I config.yml
file.
.RE
.PP
.BR "-l", " --last"
.RS 3
Display the dates of the last successful publish and promotions.
.RE
.PP
.BR "-d", " --dryrun"
.RS 3
Do not perform the actual publish activity, but show what would be done.
This option can be used to verify that the correct configuration options have been defined and applied before running the actual publication of content.
.RE
.PP
.BR "-c", " --comment"
.I "COMMENT"
.RS 3
Tag the published content view with a custom comment.
By default a comment indicating the content view was published via the API by the user running the script is used.
.RE
.PP
.BR "-q", " --quiet"
.RS 3
Do not display refreshing progress bars during the operation.
This option can be used when including this script in automated tasks, for example from cron.
.RE
.PP
.BR "-m", " --forcemeta"
.RS 3
Force the regeneration of repository metadata during the publish.
.RE


.SH EXAMPLES
Check when the last publish was performed:
.RS 6
.B "publish_content_views -l"
.RE

Publish all content views defined in the Satellite:
.RS 6
.B "publish_content_views -a"
.RE

Show what will be published by default, but do not actually publish the content:
.RS 6
.B "publish_content_views -d"
.RE

.SH FILES
.B Main Configuration
.RS 3
.I /usr/share/sat6_scripts/config/config.yml
.RE
.LP
.B Log Files
.RS 3
.I /var/log/satellite/*
.RE

.SH SEE ALSO
.BR sat6_scripts_config (8),
.BR promote_content_views (8),
.BR clean_content_views (8).

.SH AUTHOR
Geoff Gatward <ggatward@redhat.com>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [1.2.2] - 2018-02-25
### Added
- Option to tag a published content view with a custom comment
- Option to clean versions taking into account published versions
- Option to force metadata regeneration during publish and promote

### Fixed
- Fix unreferenced var in check_disk_space function
- Various minor README corrections
- auto_content did not publish CV if the satellite is internet connected
- publish_content_views KeyError exception if API returns null value
- Fix handling empty lists for used and unused versions in clean_content_views


## [1.2.1] - 2017-12-11
### Fixed
- Incorrect mapping of days in auto_content.py
- Support for username:token authentication to Artifactory Puppet Forge server
- Email now supports multiple recipients


## [1.2.0] - 2017-12-10
### Added
- push_puppetforge now supports jFrog Artifiactory repository via HTTP POST
- sat_import now checks for exports that have not been imported (missed/skipped)
- sat_import --fixhistory option to force align import/export histories
- Email notification capability for use when automating content scripts
- Add unattended option to allow scripts to be automated
- auto_content scripts to allow unattended import/publish/promote/clean activity

### Changed
- --notar export saved in /cdn_export dir rather than /export to prevent it being deleted

### Removed
- Skip GPG short option (-n)


## [1.1.1] - 2017-10-25
### Added
- Allow limiting of Publish and Promotion to specified number of CV's at once.


## [1.1.0] - 2017-10-23
### Added
- Proxy support for download_manifest
- Quiet option in publish/promote scripts
- Full repo package comparison written to log during disconnected import
- Provide a history of successful content exports and imports
- Present a warning if import has already been performed for given dataset
- Option to leave export in unarchived format (--notar)
- Option to force a disconnected (import only) Satellite to export if required

### Changed
- Updated exit codes throughout the scripts
- Updated export dataset name format to allow for multiple exports per day
- Updated to Semantic Versioning

### Fixed
- Various minor bugfixes and typos
- Fixed bug in DoV export

### Deprecated
- Skip GPG short option (-n) will be removed in 1.2.0. Long option (--nogpg) will remain.


## [1.0] - 2017-03-06
- Initial 'public' release

## 0.6 - 2017-02-27
- Last of a series of pre-release betas

[Unreleased]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.2...HEAD
[1.2.2]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.1...1.2.2
[1.2.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.0...1.1.0
[1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/0.6...1.0


publish_content_views.py
#!/usr/bin/python
#title           :publish_content_view.py
#description     :Publishes new versions of Satellite 6 content views
#URL             :https://github.com/ggatward/sat6_scripts
#author          :Geoff Gatward <ggatward@redhat.com>
#notes           :This script is NOT SUPPORTED by Red Hat Global Support Services.
#license         :GPLv3
#==============================================================================
"""
Publishes new content view versions to the Library

"""
#pylint: disable-msg=R0912,R0913,R0914,R0915

import sys
import os
import argparse
import datetime
import pickle
import simplejson as json
import helpers


def get_cv(org_id, publish_list):
    """Get the content views"""

    # Query API to get all content views for our org
    cvs = helpers.get_json(
        helpers.KATELLO_API + "organizations/" + str(org_id) + "/content_views/")
    ver_list = {}
    ver_descr = {}
    ver_version = {}

    for cv_result in cvs['results']:
        # We will never publish the DOV
        if cv_result['name'] != "Default Organization View":

            # Handle specific includes and excludes
            if publish_list and cv_result['name'] not in publish_list:
                msg = "Skipping content view '" + cv_result['name'] + "'"
                helpers.log_msg(msg, 'DEBUG')
                continue

            # Get the ID of each Content View
            msg = "Processing content view '" + cv_result['name'] + "' " + str(cv_result['id'])
            helpers.log_msg(msg, 'DEBUG')

            # Find the next version of the view
            ver_list[cv_result['id']] = cv_result['id']
            ver_descr[cv_result['id']] = cv_result['name']
            ver_version[cv_result['id']] = cv_result['next_version']

    return ver_list, ver_descr, ver_version


def publish(ver_list, ver_descr, ver_version, dry_run, runuser, description, quiet, forcemeta):
    """Publish Content View"""

    # Set the task name to be displayed in the task monitoring stage
    task_name = "Publish content view to Library"

    # Now we have all the info needed, we can actually trigger the publish.
    task_list = []
    ref_list = {}

    # Catch scenario that no CV versions are found matching publish criteria
    if not ver_list:
        msg = "No content view versions found matching publication criteria"
        helpers.log_msg(msg, 'ERROR')
        if helpers.MAILOUT:
            helpers.tf.seek(0)
            output = "{}".format(helpers.tf.read())
            helpers.mailout(helpers.MAILSUBJ_FP, output)
        sys.exit(1)

    # Break repos to publish into batches as configured in config.yml
    cvchunks = [ ver_list.keys()[i:i+helpers.PUBLISHBATCH] for i in range(0, len(ver_list), helpers.PUBLISHBATCH) ]

    # Loop through the smaller subsets of repo id's
    for chunk in cvchunks:
        for cvid in chunk:

            # Check if there is a publish/promote already running on this content view
            locked = helpers.check_running_publish(ver_list[cvid], ver_descr[cvid])

            if not locked:
                msg = "Publishing '" + str(ver_descr[cvid]) + "' Version " + str(ver_version[cvid]) + ".0"
                helpers.log_msg(msg, 'INFO')
                print helpers.HEADER + msg + helpers.ENDC

            if not dry_run and not locked:
                try:
                    task_id = helpers.post_json(
                        helpers.KATELLO_API + "content_views/" + str(ver_list[cvid]) +\
                        "/publish", json.dumps(
                            {
                                "description": description,
                                "force_yum_metadata_regeneration": str(forcemeta)
                            }
                            ))["id"]
                except Warning:
                    msg = "Failed to initiate publication of " + str(ver_descr[cvid])
                    helpers.log_msg(msg, 'WARNING')
                except KeyError:
                    msg = "Failed to initiate publication of " + str(ver_descr[cvid])
                    helpers.log_msg(msg, 'WARNING')
                else:
                    task_list.append(task_id)
                    ref_list[task_id] = ver_descr[cvid]

        # Notify user in the case of a dry-run
        if dry_run:
            msg = "Dry run - not actually performing publish"
            helpers.log_msg(msg, 'WARNING')
        else:
            # Wait for the tasks to finish
            helpers.watch_tasks(task_list, ref_list, task_name, quiet)

    # Exit in the case of a dry-run
    if dry_run:
        sys.exit(2)
    else:
        return


def main(args):
    """
    Main routine
    """

    # Who is running this script?
    runuser = helpers.who_is_running()

    # Set the base dir of the script and where the var data is
    global dir
    global vardir
    dir = os.path.dirname(__file__)
    vardir = os.path.join(dir, 'var')
    confdir = os.path.join(dir, 'config')

    # Check for sane input
    parser = argparse.ArgumentParser(
        description='Publishes content views for specified organization.')
    group = parser.add_mutually_exclusive_group()
    # pylint: disable=bad-continuation
    parser.add_argument('-o', '--org', help='Organization (Uses default if not specified)',
        required=False)
    group.add_argument('-a', '--all', help='Publish ALL content views', required=False,
        action="store_true")
    parser.add_argument('-d', '--dryrun', help='Dry Run - Only show what will be published',
        required=False, action="store_true")
    parser.add_argument('-l', '--last', help='Display last promotions', required=False,
        action="store_true")
    parser.add_argument('-c', '--comment', help="Add a custom description", required=False)
    parser.add_argument('-q', '--quiet', help="Suppress progress output updates", required=False,
        action="store_true")
    parser.add_argument('-m', '--forcemeta', help="Force metadata regeneration", required=False,
        action="store_true")

    args = parser.parse_args()

    # Log the fact we are starting
    if not args.last:
        msg = "-------- Content view publish started by " + runuser + " -----------"
        helpers.log_msg(msg, 'INFO')

    # Set our script variables from the input args
    if args.org:
        org_name = args.org
    else:
        org_name = helpers.ORG_NAME
    dry_run = args.dryrun

    # Set up the description that will be added to the published version
    if args.comment:
        description = args.comment
    else:
        description = "Published by " + runuser + "\n via API script"

    # Load the promotion history
    if not os.path.exists(vardir + '/promotions.pkl'):
        if not os.path.exists(vardir):
            os.makedirs(vardir)
        phistory = {}
    else:
        phistory = pickle.load(open(vardir + '/promotions.pkl', 'rb'))

    # Read the promotion history if --last requested
    if args.last:
        if phistory:
            print 'Last promotions:'
            for lenv, time in phistory.iteritems():
                print lenv, time
        else:
            print 'No promotions recorded'
        sys.exit(0)


    publish_list = []
    if not args.all:
        publish_list = helpers.CONFIG['publish']['content_views']

        if not publish_list:
            msg = "Cannot find publish configuration"
            helpers.log_msg(msg, 'ERROR')
            if helpers.MAILOUT:
                helpers.tf.seek(0)
                output = "{}".format(helpers.tf.read())
                helpers.mailout(helpers.MAILSUBJ_FP, output)
            sys.exit(1)

        msg = "Config found for CV's " + str(publish_list)
        helpers.log_msg(msg, 'DEBUG')

    # Get the org_id (Validates our connection to the API)
    org_id = helpers.get_org_id(org_name)

    # Get the list of Content Views along with the latest view version in each environment
    (ver_list, ver_descr, ver_version) = get_cv(org_id, publish_list)

    # Publish the content views. Returns a list of task IDs.
    publish(ver_list, ver_descr, ver_version, dry_run, runuser, description, args.quiet,
        args.forcemeta)

    # Add/Update the promotion history dictionary so we can check when we last promoted
    phistory['Library'] = datetime.datetime.strftime(datetime.datetime.now(), '%Y-%m-%d')
    pickle.dump(phistory, open(vardir + '/promotions.pkl', 'wb'))

    # Run the mailout
    if helpers.MAILOUT:
        helpers.tf.seek(0)
        output = "{}".format(helpers.tf.read())
        message = "Publish completed successfully\n\n" + output
        subject = "Satellite 6 publish completed"
        helpers.mailout(subject, message)

    # Exit cleanly
    sys.exit(0)

if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except KeyboardInterrupt, e:
        print >> sys.stderr, ("\n\nExiting on user cancel.")
        sys.exit(1)
